Replace `Commands` with `Service` and `ServiceStore`

sebastianvera opened the pull request:
We needed a way to associate `commands` with `repos`, so I introduced the concept of a Service that encapsulates both structs, and knows how to start, stop and update itself.

I also added a `ServiceStore` struct, that is a `map[string]*Service` where the key is the service's name.

I didn't like too much the idea of representing a command and a repo relation by the command's slug, so now all services have a `name` attribute, that identifies them in the store and it should be the same as its command's slug.

I also updated the realtime package, to publish the new defined payload to the healthz topic:

```json
{
  "data": [
    {
      "name": "wisebot-core",
      "version": "1a2b3c4d",
      "status": "running",
      "repo_version": "1a2b3c4d"
    }
  ]
}
```

sebastianvera committed:
Replace `Commands` with `Service` and `ServiceStore`

sebastianvera committed:
Remove logger, slug and add Clone function to `command`

sebastianvera committed:
Add `command.Clone` and add logs

sebastianvera committed:
Update main by removing command slug

sebastianvera committed:
Clone repo to given path and fix comments

sebastianvera committed:
Update repo path

mtavano reviewed (commented):
Looks good to me. Maybe, would be great make a review together. It is possible that we have overlooked something

jaguirresza commented on services.go:
```diff
@@ -0,0 +1,167 @@
+// ServiceStore represents a set of commands.
+// It has convinient methods to run and stop all
+// commands.
+type ServiceStore map[string]*Service
```
I think this should be protected by a mutex since is updated concurrently.

sebastianvera replied:
I totally missed it, I'll add it

jaguirresza commented on services.go:
```diff
@@ -0,0 +1,162 @@
+
+// Start starts all the commands inside the command list by
+// looping and calling each command Start function.
+func (ss *ServiceStore) Start() error {
```
In a case of error what happens? Should we try at least to stop started services?

jaguirresza replied:
Also maybe is a good idea to known if a service is already running to not start a started command.
Or is already implemented that way?

sebastianvera replied:
I'm not sure, it can get a little messy, for example, if the application already started and this functions got called when the process was updating, we should not stop other processes, but if the app is just booting, we should stop other process. Maybe that's the approach we need to follow

sebastianvera replied:
To answer your second question, we are not checking if the process was already started, but if you tried tu run the process, it will throw an error either way ... we should add that validation.

jaguirresza replied:
Related to the first point, maybe we should ignore errors and try to start as many processes as possible and just marked failed ones with a status `StatusError`. But is seems kind of arbitrary start all the process until one processes fail.
I think we should define a predictable behaviour.

jaguirresza replied:
Maybe we can add one more method `MustStart` that panics in the case of a failure and `Start` don't return an error at all.

sebastianvera replied:
I agree, but if one process fails to start when the operator is booting up, maybe is better approach to no start at all, since is crucial to have al processes running. On the other hand, if a process, after updating, fails to start, is not THAT critical, but it still critical (if you know what I mean), maybe, in that case, we can try to perform a rollback or are least try to fix it

sebastianvera replied:
Definetly, I'll add that method, that'd resolve our booting problem

sebastianvera replied:
But if it panics, we should recover from that panic, and stop the already running processes and continue with the gracefully shutdown

jaguirresza replied:
Ahh, that's true, then return an error and try to stop all process should be a valid approach.

sebastianvera replied:
I need to check what happens when I try to stop a non-running process. Maybe we can just omit that error (that's how is programmed right now)

jaguirresza commented on command/command.go:
```diff
@@ -82,34 +100,30 @@ func (c *Command) Update() (updated bool, err error) {
 		return c.status
 	}
 
 	ps := c.Cmd.ProcessState
```
maybe is clearer the intention written like this:

```go
func (c *Command) Status() Status {
    if ps := c.Cmd.ProcessState; ps != nil && ps.Exited() {
        if ps.Suceess() {
            return StatusDone
        }
        return StatusError
    }
    return c.status
}
```

sebastianvera replied:
Yes, the truth is I didn't know how to get the correct status, and after debugging and fixing the code it ended up like that, I didn't even try to refactor it, but it always should look easier, like your approach.

jaguirresza commented on command/command.go:
```diff
@@ -7,41 +7,50 @@ import (
-
-	Slug    string
 	Version string
+	Updater Updater
```
should be in lower since `Command.Updater.Update` method should be called though `Command.Update`

jaguirresza replied:
Ohh I see the problem now. The construtor is the problem.
What about drop `Updater` field and change `Updater` type from an interface to a function and receive an `Updater` in the Update method?

```go

type Updater func (*Command) (bool, error)

func (c *Command) Update(updater Updater) (bool, error)
```

sebastianvera replied:
Love it!

sebastianvera replied:
I didn't change the Updater interface, since it'd affect the current design

sebastianvera committed:
Remove Updater from command struct

Instead of storing the Updater in the struct,
receive it from parameter in the `Update` function.

sebastianvera committed:
Simplify `command.Status` function

sebastianvera committed:
Add RWMutex to ServiceStore

sebastianvera committed:
Update main to use new ServiceStore

sebastianvera committed:
Stop process if start fails when booting operator

jaguirresza commented on services.go:
```diff
@@ -44,13 +45,18 @@ func (s *Service) Update() (bool, error) {
-	for _, svc := range ss {
+func (ss *ServiceStore) MarshalJSON() ([]byte, error) {
+	services := make([]*Service, len(ss.list))
+	for _, svc := range ss.list {
```
the critical section is `ss.list` so all `for` loop should be protected.

jaguirresza commented on services.go:
```diff
@@ -109,16 +125,23 @@ func (ss *ServiceStore) Update(name string) error {
 	s := newService(name, c, r)
-	(*ss)[s.Name] = s
+
+	if ss.list == nil {
```
all `ss.list` operation should be protected

jaguirresza commented on services.go:
```diff
@@ -130,7 +153,7 @@ func (ss *ServiceStore) StartService(name string) error {
 // looping and calling each command Start function.
 func (ss *ServiceStore) Start() error {
-	for _, svc := range *ss {
+	for _, svc := range ss.list {
```
the critical section is ss.list so all for loop should be protected.

jaguirresza commented on services.go:
```diff
@@ -155,7 +178,7 @@ func (ss *ServiceStore) StopService(name string) error {
 // looping and calling each service command's Stop function.
 func (ss *ServiceStore) Stop() error {
-	for _, svc := range *ss {
+	for _, svc := range ss.list {
```
the critical section is ss.list so all for loop should be protected.

sebastianvera committed:
Add missing mutexes on critical sections

jaguirresza commented on services.go:
```diff
@@ -129,8 +129,15 @@ func (ss *ServiceStore) Update(name string) error {
-	if ss.list == nil {
+	// Is this correct?
+	ss.mu.RLock()
+	list := ss.list
```
i think that this is enough:

```go
func (ss *ServiceStore) Save(name string, c *command.Command, r *git.Repo) {
	s := newService(name, c, r)

        ss.mu.Lock()
	defer ss.mu.Unlock()

	if ss.list  == nil {
		ss.list = make(map[string]*Service)
	}
	ss.list[s.Name] = s
}
```

sebastianvera committed:
Fix what @jaguirre commented
